ouroboros-network/test/CDDL.hs

```diff
@@ -3,134 +3,215 @@
-import qualified Codec.CBOR.Decoding as CBOR
-import qualified Codec.CBOR.Term as CBOR (Term)
+import Codec.CBOR.Decoding as CBOR (Decoder, decodeBytes ,decodeListLenOf, decodeWord)
+import Codec.CBOR.Encoding (encodeBytes, encodeListLen, encodeWord)
+import Codec.CBOR.Term as CBOR
 import Codec.CBOR.Read
+import Codec.CBOR.Write (toLazyByteString)
+
+import Ouroboros.Network.Block (BlockNo, HeaderHash, Point)
+import Ouroboros.Network.Testing.ConcreteBlock (BlockHeader (..), Block)
 
-import Ouroboros.Network.Block (Point)
 import Ouroboros.Network.Protocol.ChainSync.Type as CS
 import Ouroboros.Network.Protocol.ChainSync.Codec (codecChainSync)
-import Network.TypedProtocol.ReqResp.Type as ReqResp
-import Network.TypedProtocol.ReqResp.Codec.Cbor (codecReqResp)
-import Network.TypedProtocol.PingPong.Type as PingPong
-import Network.TypedProtocol.PingPong.Codec.Cbor (codecPingPong)
+import Ouroboros.Network.Protocol.ChainSync.Test ()
 import Ouroboros.Network.Protocol.BlockFetch.Codec (codecBlockFetch)
+import Ouroboros.Network.Protocol.BlockFetch.Test ()
 import Ouroboros.Network.Protocol.BlockFetch.Type as BlockFetch
 import Ouroboros.Network.Protocol.Handshake.Codec (codecHandshake)
 import Ouroboros.Network.Protocol.Handshake.Type as Handshake
-import Ouroboros.Network.Block (HeaderHash)
+import Ouroboros.Network.Protocol.Handshake.Test (VersionNumber)
+import Ouroboros.Network.Protocol.TxSubmission.Codec (codecTxSubmission)
+import Ouroboros.Network.Protocol.TxSubmission.Type as TxSubmission
+import Ouroboros.Network.Protocol.TxSubmission.Test (TxId, Tx)
 
 import Network.TypedProtocol.Codec
+import Network.TypedProtocol.ReqResp.Codec.Cbor (codecReqResp)
+import Network.TypedProtocol.ReqResp.Type as ReqResp
+import Network.TypedProtocol.PingPong.Type as PingPong
+import Network.TypedProtocol.PingPong.Codec.Cbor (codecPingPong)
 
--- the concrete types used for the test
-type CS = ChainSync DummyBytes (Point DummyBytes)
-type RR = ReqResp DummyBytes DummyBytes
-type BF = BlockFetch DummyBytes
-type HS = Handshake Word32 CBOR.Term
+-- These files must be in place:
+specFile :: FilePath
+specFile = "test/messages.cddl"
+
+cddlTool :: FilePath
+cddlTool = "cddl"
+
+diag2cborTool :: FilePath
+diag2cborTool = "diag2cbor.rb"
 
 main :: IO ()
-main = generateAndDecode "cddl" "diag2cbor.rb" "test/messages.cddl" 100
+main = defaultMain tests
+
+tests :: TestTree
+tests =
+  adjustOption (const $ QuickCheckMaxSize 20) $ testGroup "messages.cddl-spec"
+  [
+-- These tests call the CDDL-tool to parse an arbitray message.
+-- The parser of the CDDL-tool is slow (exponential runtime & space).
+    testProperty "encode ChainSync"             prop_specCS
+  , testProperty "encode Request-Response"      prop_specReqResp
+  , testProperty "encode BlockFetch"            prop_specBF
+  , testProperty "encode TxSubmission"          prop_specTxSubmission
+  , testProperty "encode Handshake"             prop_specHandshake
+
```

Comment:
It looks like you're not testing the Ping-Pong protocol, is that intentional?

Reply:
I added PingPong. please see the the commits